### PR TITLE
feat: Token statistics tooltip for each API request

### LINKED ISSUE #7740: Add detailed token stats for every API call
### What specific problem does this solve?

Screenshot from 2025-09-06 18-11-28

Currently, each API request in the statistics only shows the price. However, many models are now free or have a low cost, making these figures less informative.

I believe it would be very convenient to optimize prompts and requests if not only the price but also the number of tokens for each request were visible. The number of cached tokens would be especially valuable.

Optionally, it would also be interesting to see the speed of incoming and outgoing tokens, although this is somewhat less important.

Assuming that some users may find this information excessive, I would also add an option to hide this statistics (just in case).

I could potentially try to implement this improvement myself if the maintainers agree to accept this feature.

### Additional context (optional)

There is some html example how it could be done from dev tools, that could help to implement the view.

``<div style="display: flex; align-items: center; gap: 10px; flex-grow: 1;"><div style="width: 16px; height: 16px; display: flex; align-items: center; justify-content: center;"><span class="codicon codicon-check" style="color: var(--vscode-charts-green); font-size: 16px; margin-bottom: -1.5px;"></span></div><span style="color: var(--vscode-foreground); font-weight: bold;">API Request</span><vscode-badge circular="" class="circular" style="opacity: 1;">$0.0011</vscode-badge><div class="flex items-center gap-1 flex-wrap"><span>↑ 24.8k (20k cache)</span><span>↓ 177</span></div></div>``

### Roo Code Task Links (Optional)

_No response_

### Request checklist

- [x] I've searched existing Issues and Discussions for duplicates
- [x] This describes a specific problem with clear impact and context

### Interested in implementing this?

- [x] Yes, I'd like to help implement this feature

### Implementation requirements

- [x] I understand this needs approval before implementation begins

### How should this be solved? (REQUIRED if contributing, optional otherwise)

See the screenshot above.

### How will we know it works? (Acceptance Criteria - REQUIRED if contributing, optional otherwise)

Given I have a task for the AI
When I run it
Then API requests occur
And in addition to the price of each request, I also see the number of input tokens (in parentheses cache hit) and the number of output tokens (numbers in k - thousands of tokens)
And this happens with every API request as the AI executes the task

### Technical considerations (REQUIRED if contributing, optional otherwise)

Adjust the user interface by showing details that the provider supplies with every request (see the screenshot above). Some calculations might be required.

Adding an option to disable the output makes the task significantly more complex, so I would include this in the next iteration (depending on user feedback?)

### Trade-offs and risks (REQUIRED if contributing, optional otherwise)

- Some users may find this information excessive
- It may look worse on a very narrow window

### PR DESCRIPTION
**Pull request by mikhailsal**
This PR replaces the previous attempt, which had to be abandoned due to an AI error (context loss). See the prior PR: https://github.com/RooCodeInc/Roo-Code/pull/7741

What’s included (difference vs main, in brief):
- Token usage statistics per API request shown in a tooltip (input, output, cache reads).
- Counts are formatted (e.g. "1.2k") and localized labels are used.
- Tooltip is shown even when there is no price, so token info remains accessible.

Rationale:
- Makes token usage visible for every request without cluttering the UI.
- Keeps details available on hover while preserving compactness.

Fixes #7740

----

> [!IMPORTANT]
> Adds a tooltip for token statistics in API requests, updates localization, and introduces token formatting utilities.
> 
>   - **Behavior**:
>     - Adds token usage statistics tooltip in `ChatRow.tsx` for API requests, displaying input, output, and cache reads.
>     - Tooltip is visible even without price information.
>   - **Utilities**:
>     - Introduces `formatTokenCount()` and `formatTokenStats()` in `formatTokens.ts` for formatting token data.
>   - **Localization**:
>     - Updates localization files (`chat.json`) for multiple languages to include token statistics labels.
>   - **Styles**:
>     - Adds CSS rules in `index.css` to hide API request text on narrow screens.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for b9b8dd524c4bf40e15bf0694fec5fb969333c649. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: webview-ui/src/utils/formatTokens.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		inputDisplay = `${inputDisplay} (${cacheDisplay} cache)`
+	}
+
+	const outputDisplay = formatTokenCount(tokensOut)
+
+	return {
+		input: inputDisplay,
+		output: outputDisplay,
+	}
+}
```
Comment (ghost) [inline webview-ui/src/utils/formatTokens.ts:42]: Consider adding unit tests for these formatting functions. They handle several edge cases (undefined, 0, < 1000, < 10k, >= 10k) that would benefit from test coverage to ensure the formatting logic works correctly.

You could create a test file at  with tests for both functions.

File: webview-ui/src/index.css [hunk 2]
```diff
@@ -486,4 +486,24 @@ input[cmdk-input]:focus {
 	transition-property: color, background-color, border-color, text-decoration-color, fill, stroke;
 	transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
 	transition-duration: 150ms;
+
```
Comment (ghost) [inline webview-ui/src/index.css:493]: Is this intentional? The media queries and container queries (lines 490-509) appear to be incorrectly nested inside the  rule. These responsive styles should probably be at the root level of the CSS file, not inside a transition utility class.

File: webview-ui/src/components/chat/ChatRow.tsx [hunk 3]
```diff
@@ -1111,13 +1122,65 @@ export const ChatRowContent = ({
...
+											overflow: "hidden",
+											textOverflow: "ellipsis",
+											flexShrink: 1,
+											minWidth: 0,
+										}}>
+										{title}
+									</span>
+									<div style={{ display: "flex", alignItems: "center", gap: "8px", flexShrink: 0 }}>
+										{hasTokenData ? (
+											<StandardTooltip
```
Comment (ghost) [inline webview-ui/src/components/chat/ChatRow.tsx]: Have you considered the accessibility of this tooltip for keyboard users? The StandardTooltip component might already handle keyboard accessibility, but it's worth verifying that users can access token statistics without a mouse.

File: webview-ui/src/utils/formatTokens.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		return `${Math.round(thousands)}k`
+	}
+}
+
+/**
+ * Format token statistics for display
+ * @param tokensIn - Input tokens
+ * @param tokensOut - Output tokens
+ * @param cacheReads - Cache read tokens (optional)
+ * @returns Formatted string for display
```
Comment (ghost) [inline webview-ui/src/utils/formatTokens.ts]: Minor documentation issue: The JSDoc says "returns Formatted string for display" but the function actually returns an object with  and  properties. Consider updating to:

**Comment by mikhailsal** [comment]
I've done. Latest screenshots:
Screenshot from 2025-09-12 12-01-23
Screenshot from 2025-09-12 12-01-02

File: webview-ui/src/utils/formatTokens.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
+/**
+ * Format token count for display
+ * @param count - The token count to format
+ * @returns Formatted string (e.g., "1.2k" for 1200)
+ */
+export function formatTokenCount(count: number | undefined): string {
```
Comment (daniel-lxs) [inline webview-ui/src/utils/formatTokens.ts:8]: Consider reusing the existing  utility from  instead of creating a duplicate implementation. The existing utility already handles number formatting with proper localization support through i18next.

File: webview-ui/src/utils/formatTokens.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+		inputDisplay = `${inputDisplay} (${cacheDisplay} ${cacheLabel})`
+	}
+
+	const outputDisplay = formatTokenCount(tokensOut)
+
+	return {
+		input: inputDisplay,
+		output: outputDisplay,
+	}
+}
```
Comment (daniel-lxs) [inline webview-ui/src/utils/formatTokens.ts:42]: This file is missing unit tests. Since similar utilities like  have comprehensive test coverage in , would you consider adding tests for these formatting functions to ensure they work correctly with various edge cases?

**Comment by mikhailsal** [comment]
@daniel-lxs I made corrections based on your feedback. We are now reusing code, and there is a unit test for the new code.
